File: leetcode/h100/group_1/code14/merge.go
package code14

/*以数组 intervals 表示若干个区间的集合，其中单个区间为 intervals[i] = [starti, endi] 。请你合并所有重叠的区间，并返回 一个不重叠的区间数组，该数组需恰好覆盖输入中的所有区间 。*/

func merge(intervals [][]int) [][]int {
	if len(intervals) == 0 {
		return [][]int{}
	}
	ans := [][]int{}
	maxEnd := 0
	for _, interval := range intervals {
		maxEnd = max(maxEnd, interval[1])
	}
	diff := make([]int, maxEnd*2+2)
	for _, interval := range intervals {
		diff[interval[0]*2]++
		diff[interval[1]*2+1]--
	}
	presum := make([]int, len(diff))
	presum[0] = diff[0]
	for i := 1; i < len(diff); i++ {
		presum[i] = presum[i-1] + diff[i]
	}
	start := -1
	for i, cnt := range presum {
		if cnt > 0 {
			if start == -1 {
				start = i
			}
		} else if start != -1 {
			ans = append(ans, []int{start / 2, i / 2})
			start = -1
		}
	}
	return ans
}

File: leetcode/hot100-java/group_1/merge.java
/*
 * 56. 合并区间
 *
 * 给定区间数组 intervals，其中 intervals[i] = [start_i, end_i]。合并所有
 * 相互重叠的区间，返回一组互不重叠且恰好覆盖原有全部区间的新区间。
 *
 * 算法实现说明：
 * 1. 先按区间起点升序排序，使可能重叠的区间在数组中连续出现。
 * 2. start、end 保存当前正在合并的区间；若下一区间起点不超过 end，就扩大 end 覆盖它。
 * 3. 若下一区间与当前区间分离，就把当前区间加入结果，并从下一区间开始新的一段。
 * 4. 扫描结束后补入最后一段，得到的区间按序且互不重叠，并覆盖全部输入区间。
 *
 * 时间复杂度：O(n log n)，主要来自排序；除返回结果和排序栈外，空间复杂度：O(log n)。
 */
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

class Solution {
    public int[][] merge(int[][] intervals) {
        if (intervals.length == 0) {
            return new int[0][];
        }
        Arrays.sort(intervals, (a, b) -> Integer.compare(a[0], b[0]));

        List<int[]> merged = new ArrayList<>();
        int start = intervals[0][0];
        int end = intervals[0][1];
        for (int i = 1; i < intervals.length; i++) {
            if (intervals[i][0] <= end) {
                end = Math.max(end, intervals[i][1]);
            } else {
                merged.add(new int[] {start, end});
                start = intervals[i][0];
                end = intervals[i][1];
            }
        }
        merged.add(new int[] {start, end});
        return merged.toArray(new int[merged.size()][]);
    }
}
